anndata/tests/test_awkward.py

```diff
@@ -0,0 +1,294 @@
+    adata = AnnData()
+    adata.uns["awk"] = array
+    adata_path = tmp_path / "adata.h5ad"
+    adata.write_h5ad(adata_path)
+
+    adata2 = read_h5ad(adata_path)
+
+    assert_equal(adata.uns["awk"], adata2.uns["awk"], exact=True)
+
+
+@pytest.mark.parametrize(
+    "arrays,expected",
+    [
+        [
+            [ak.Array([{"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}]), None],
+            ak.Array([{"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}, {}, {}, {}]),
+        ],
+        [
+            [None, ak.Array([{"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}])],
+            ak.Array([{}, {}, {}, {"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}]),
+        ],
+        [
+            [
+                None,
+                ak.Array([{"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}]),
+                pd.DataFrame(),
+            ],
+            ak.Array([{}, {}, {}, {"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}]),
+        ],
+        [
+            [
+                ak.Array([{"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}]),
+                pd.DataFrame(),
+            ],
+            ak.Array([{"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}]),
+        ],
+        [
+            [
+                ak.Array([{"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}]),
+                pd.DataFrame().assign(a=[3, 4], b=[5, 6]),
+            ],
+            NotImplementedError,
+        ],
+        [
+            [
+                ak.Array([{"a": [1, 2], "b": [1, 2]}, {"a": [3], "b": [4]}]),
+                np.ones((3, 2)),
+            ],
+            NotImplementedError,
+        ],
+    ],
+)
+@pytest.mark.parametrize("key", ["obsm", "varm"])
+@pytest.mark.parametrize("join", ["outer", "inner"])
+def test_concat_mixed_types(key, arrays, expected, join):
+    """Test that concatenation of AwkwardArrays with arbitrary types, but zero length dimension
+    or missing values works."""
+    axis = 0 if key == "obsm" else 1
+
+    to_concat = []
+    for a in arrays:
+        shape = np.array([3, 3])  # default shape (in case of missing array)
+        if a is not None:
+            length = dim_len(a, 0)
+            shape[axis] = length
+
+        tmp_adata = gen_adata(
+            tuple(shape), varm_types=(), obsm_types=(), layers_types=()
+        )
+
+        if a is not None:
+            if isinstance(a, pd.DataFrame):
+                a.set_index(
+                    tmp_adata.obs_names if key == "obsm" else tmp_adata.var_names,
+                    inplace=True,
+                )
+            getattr(tmp_adata, key)["test"] = a
+
+        to_concat.append(tmp_adata)
+
+    if isinstance(expected, type) and issubclass(expected, Exception):
+        with pytest.raises(expected):
+            anndata.concat(to_concat, axis=axis, join=join)
+    else:
+        result = anndata.concat(to_concat, axis=axis, join=join)
+        assert_equal(getattr(result, key)["test"], expected, exact=True)
```

Comment:
@grst, let's push this feature to a future PR?

Reply:
Sure! maybe we can already keep the testcase somewhere? Either in a draft PR, or just set it to `xfail`?

Reply:
Pushed a commit that makes more of these not implemented errors. That will keep them around, but we've still got the cases recorded.

I realized the "simple" empty cases are kinda important. E.g. concat where one of the entries doesn't have anything under that key.